docs/Plugins.rst

```diff
@@ -1159,6 +1159,59 @@ Examples:
 ``autolabor CUTWOOD disable``
         Turn off autolabor for wood cutting.
 
+.. _channel-safely:
+
+channel-safely
+==============
+Automatically manage channel designations to safely complete them.
+Channel-safely keeps dwarves safe by marking unsafe designations for
+later. A designation is unsafe if it is under another designation or
+over empty space. The latter situation will never be safe, but with
```

Comment:
It's considered unsafe because the space below is empty, so when they dig themselves down they fall into the empty. If there are ways to prevent this it can't be easily ensured afaik.

I'm gonna go and add some comments throughout the source when I find some time.. but from what I remember "over empty space" essentially means they channeled into a channel or cavern, so there won't be a floor directly below them to fall onto which might already hurt them.. like bruise their feet.. or ego.. or something.

Reply:
That's what I don't get, though - dwarves don't stand on the tile they're channeling unless they approaching it from the tile directly above it. They normally stand to the side as they channel, don't they?

Reply:
I don't think so, I can double check. Maybe you're thinking of situations where they can't stand on the tile they're going to channel so they have to dig it first then channel? Or I could be mistaken, but ime they fall if there is nothing below the tile so either way it'd be unsafe cause it's not worth checking where they are standing in my opinion

Reply:
Glad I double checked. Guess I gotta eat my words. Dunno if I thought that because I had been seeing an edge case or what. Not sure I want to revise it to path find a path out.. but I suppose people would like that =/

Reply:
@myk002 so how would I detect this true unsafe over empty space case? (from a high level)

Reply:
Ugh.. I had written a whole train of thought on my phone but didn't submit it and now it's gone. Let me see what I can reconstruct.

A lot of this will be duplicating what you already have, but let me see if I can form a chain of reasoning.

1. digging a channel is unsafe if it is likely to cause a dwarf to fall and get injured
2. a dwarf can only fall if there is empty space under the tile that is being channeled and a dwarf is on top of the tile when it is channeled
3. situations where a dwarf will be on top of the channeled tile:
    a) dwarf is channeling that tile and can only approach that tile from above OR was already standing on that tile when they took the job to dig it
    b) two channel designations are adjacent and a dwarf is standing on one designated tile while digging the other
    c) dwarf is pathing through the tile on the way to somewhere else
4. digging a channel is also undesirable (though not necessarily unsafe to any dwarves) if after digging the channel other designated tiles are no longer reachable

So, potential algorithm.

1. find groups of connected channeling designations (a "project)
2. put all designations in the project in marker mode that have project tiles directly above them (so when those tiles are channeled we don't have new empty space to contend with)
3. for unmarkered tiles that are over open space, marker them all, then unmarker a tile whose removal would not make any adjacent tile unreachable AND whose path does not involve a dwarf coming straight down at it. You can probably do more than one of these tiles at a time, but we'd have to think carefully about how.

also, we can set all passable tiles that are being channeled as a restricted traffic area to minimize other dwarves passing through at a bad time

test cases:

- channeling 10 z-levels straight down through rock should work. the tiles should be channeled strictly top to bottom, even if the tiles on middle levels are accessible to other miners
- channeling a 10x10 floor area over magma should slowly dig out the tiles one by one until they are all clear
- channeling 3 z-levels straight down through the ceiling of a cavern should stop at the last tile before the ceiling of the cavern is breached (since the path to that tile is straight down). The dwarf will become stuck.

how to calculate a path? df has a cache you can use, but how to test changes to that pathing? This I don't know.

Reply:
> 2. ...on top of the tile when

This made me think of an easy partial solution whereby when a job is started the plugin spawns a tick event, probably a customized lambda, to check if the job's tile is occupied by a dwarf until the event handler for ticks is unregistered when the completed job event fires. During a tick even, when the job tile is occupied we pause the job somehow.. or cancel and requeue it(?).

I think then we could ignore pathing entirely.



> 2. put all designations in the project in marker mode that have project tiles directly above them (so when those tiles are channeled we don't have new empty space to contend with)
> 
> 3. for unmarkered tiles that are over open space, marker them all, then unmarker a tile whose removal would not make any adjacent tile unreachable AND whose approach path does not involve a dwarf coming straight down at it. You can probably do more than one of these tiles at a time, but we'd have to think carefully about how.

I think this would be a change from what I have (or did have, at least), but I think it would be a welcome one.

What I think it currently does is effectively wait until the top-most overlapping layer of designations are finished to enable any overlapping groups below them. And changing, I don't think I'd need to worry about groups.. but I dunno maybe it's better to wait for the group to finish to ensure paths above aren't ruined.. but then we've maybe made digging pits more difficult to complete due to project lifespan and the size of the hole. I can foresee trapped dwarves potentially, who must dig to escape.. might just be a fun! addition to keep?

Reply:
I'm not sure when I'll get to this. I've been busy searching for an apartment and with work and uni. I foresee a large refactor, and perhaps reimplementation coming though